apis/projectcontour/v1alpha1/server.go

```diff
@@ -0,0 +1,72 @@
+// Copyright Project Contour Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1alpha1
+
+import (
+	"fmt"
+)
+
+// XDSServerType is the type of xDS server implementation.
+type XDSServerType string
+
+const ContourServerType XDSServerType = "contour"
+const EnvoyServerType XDSServerType = "envoy"
+
+// Validate the xDS server type.
+func (s XDSServerType) Validate() error {
+	switch s {
+	case ContourServerType, EnvoyServerType:
+		return nil
+	default:
+		return fmt.Errorf("invalid xDS server type %q", s)
+	}
+}
+
+// ServerParameters holds the config for the Contour xDS server.
+type ServerParameters struct {
+	// Defines the XDSServer to use for `contour serve`.
+	// Defaults to "contour"
+	// +kubebuilder:default=contour
+	// +kubebuilder:validation:Enum=contour;envoy
+	Type XDSServerType `json:"type,omitempty"`
+
+	// Defines the xDS gRPC API address which Contour will serve.
+	// Defaults to "0.0.0.0".
+	//  +optional
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:default="0.0.0.0"
+	Address string `json:"address,omitempty"`
```

Comment:
In fields like this where it's optional but has a default, `omitempty` will never do anything because the field can never be empty. It will always have at least the default value. So I guess I think we should take it out.